linkerd/app/core/src/control.rs

```diff
@@ -358,150 +188,148 @@ pub mod dns_resolve {
         }
 
         fn call(&mut self, target: ControlAddr) -> Self::Future {
-            Box::pin(futures::future::ok(match target.addr {
-                Addr::Socket(sa) => Resolution {
-                    current_targets: HashSet::default(),
-                    address: target.clone(),
-                    dns: self.dns.clone(),
-                    state: State::Constant(Some(Target {
-                        addr: sa,
-                        server_name: target.identity.clone(),
-                    })),
-                },
-                Addr::Name(ref na) => Resolution {
-                    current_targets: HashSet::default(),
-                    address: target.clone(),
-                    dns: self.dns.clone(),
-                    state: State::Resolving(self.dns.resolve_ips(na.name())),
-                },
-            }))
+            Box::pin(futures::future::ok(from_stream::<UpdatesStream>(Box::pin(
+                resolution_stream(self.dns.clone(), target.clone()),
+            ))))
         }
     }
 
-    impl Resolution {
-        pub fn diff_targets(
-            old_targets: &HashSet<Target>,
-            new_targets: &HashSet<Target>,
-        ) -> (Vec<(SocketAddr, Target)>, Vec<SocketAddr>) {
-            let mut adds = Vec::default();
-            let mut removes = Vec::default();
-
-            for new_target in new_targets.iter() {
-                if !old_targets.contains(new_target) {
-                    adds.push((new_target.addr, new_target.clone()))
-                }
-            }
-
-            for old_target in old_targets.iter() {
-                if !new_targets.contains(old_target) {
-                    removes.push(old_target.addr)
-                }
-            }
+    // === impl ControlPlaneResolution ===
 
-            (adds, removes)
-        }
-    }
+    use async_stream::try_stream;
+    use futures::prelude::*;
+    use tokio::time::Instant;
 
-    // === impl ControlPlaneResolution ===
+    pub fn diff_targets(
+        old_targets: &HashSet<Target>,
+        new_targets: &HashSet<Target>,
+    ) -> (Vec<(SocketAddr, Target)>, Vec<SocketAddr>) {
+        let mut adds = Vec::default();
+        let mut removes = Vec::default();
 
-    impl resolve::Resolution for Resolution {
-        type Endpoint = Target;
-        type Error = Error;
+        for new_target in new_targets.iter() {
+            if !old_targets.contains(new_target) {
+                adds.push((new_target.addr, new_target.clone()))
+            }
+        }
```

Comment:
I know this code was already there, so don't worry about changing it if you don't want to, but I think we could probably do this with [`HashSet::difference`](https://doc.rust-lang.org/std/collections/struct.HashSet.html#method.difference).
```suggestion
        for new_target in new_targets.difference(old_targets) {
                adds.push((new_target.addr, new_target.clone()))
        }
```
or even
```suggestion
        let removes = new_targets.difference(old_targets)
            .map(|new_target| (new_target.addr, new_target.clone())
            .collect();
```